### PR TITLE
[Intel MKL] Enable pad+fusedconv fusion

### PR DESCRIPTION
**Pull request by guizili0**

### CODE REVIEW

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you very much for your PR and so sorry for my delay!

File: tensorflow/core/graph/mkl_layout_pass.cc [hunk 1]
```diff
@@ -917,6 +927,59 @@ class MklLayoutRewritePass : public GraphOptimizationPass {
...
+          break;
+        }
+      }
+    }
+    // Check if only VALID type of padding is used
+    // or not.
+    if (n != nullptr) {
+      string padding;
+      TF_CHECK_OK(GetNodeAttr(conv_node->def(), "padding", &padding));
+      if (padding != "VALID") {
```
[the remark this answers is not in the document]
Comment (guizili0) [reply tensorflow/core/graph/mkl_layout_pass.cc]: @penpornk we only merge the case that pad only have one output node _FusedConv2D.
below code did this check.
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/graph/mkl_layout_pass.cc#L2644

[the remark this answers is not in the document]
Comment (penpornk) [reply tensorflow/core/graph/mkl_layout_pass.cc]: @guizili0 Got it. Sorry I always forgot about this!

[the remark this answers is not in the document]
Comment (guizili0) [reply tensorflow/core/graph/mkl_layout_pass.cc]: @penpornk that's fine.

File: tensorflow/core/kernels/mkl_fused_ops_test.cc [hunk 2]
```diff
@@ -401,5 +403,262 @@ TEST_F(FusedPadConvOpTest, PaddingConvTestNchw) {
...
+      AddInputFromArray<uint8>(dummy_shape, dummy_tensor);
+    TF_ASSERT_OK(RunOpKernel());
+
+    // Compare output to expected results
+    const Tensor& output_tensor = *GetOutput(0);
+    // Index 2 will need to be changed if the number of outputs produced
+    // by MklConv2D change.
+    const Tensor& output_meta_tensor = *GetOutput(2);
+    ConvMklToTF<T> conv_comp;
+    conv_comp.PerformConversion(dtype, output_tensor, output_meta_tensor,
```
[the remark this answers is not in the document]
Comment (guizili0) [reply tensorflow/core/kernels/mkl_fused_ops_test.cc]: @penpornk seems the function PerformConversion cannot be static.
I got error when I did this.
./tensorflow/core/lib/core/status_test_util.h:24:3: error: cannot call member function 'tensorflow::Status tensorflow::OpsTestBase::InitOp()' without object
   EXPECT_EQ(::tensorflow::Status::OK(), (statement))
   ^
tensorflow/core/kernels/mkl_fused_ops_test.cc:57:5: note: in expansion of macro 'TF_EXPECT_OK'
     TF_EXPECT_OK(InitOp());

**Comment by guizili0** [comment]
@penpornk I have modified code based on your comments, after internal test passed, I will push the code for review, sorry about the delay.

**Comment by guizili0** [comment]
@penpornk your comments had been addressed, please help to review, sorry about this delay.

File: tensorflow/core/graph/mkl_layout_pass_test.cc [hunk 3]
```diff
@@ -1517,34 +1518,31 @@ TEST_F(MklLayoutPassTest, Output_ControlEdge_PadWithFusedConv2D_Positive) {
...
 
 // Pad + _FusedConv2D with padding is VALID,
 // Input node pointing to both Pad and _FusedConv2D
-// Output of both Pad and _FusedConv2D feeds one node (Z as Output2)
-// A = input(as image), B = input(as paddings), C= Pad
-// F = input(as bias), E = _FusedConv2D, Z = Output2
-// C=Pad(A,B); E=_FusedConv2D(C,A,F); Z=Output(C,E)
+// Output of both Pad and _FusedConv2D feeds one node (G as Output2)
+// A = input(as image), B = input(as paddings), C = Pad(A, B)
+// E = input(as bias), F = _FusedConv2D(C, A, E), G=Output(C, F)
```
Comment (penpornk) [inline tensorflow/core/graph/mkl_layout_pass_test.cc]: `G=Output` -> `G = Output`. Sorry for nitpicking!

File: tensorflow/core/kernels/mkl_fused_ops_test.cc [hunk 4]
```diff
@@ -594,10 +557,8 @@ class MklPadWithFusedConv2DOpTest : public OpsTestBase {
     for (const Tensor& arg : args)
       AddInputFromArray<T>(arg.shape(), arg.flat<T>());
     AddInputFromArray<int32>(padding.shape(), padding.flat<int32>());
-    AddInputFromArray<uint8>(dummy_shape, dummy_tensor);
-    AddInputFromArray<uint8>(dummy_shape, dummy_tensor);
-    AddInputFromArray<uint8>(dummy_shape, dummy_tensor);
-    for (const Tensor& arg : args)
+    // Add MKL meta input for input, filter, pad and agrs.
+    for (int i = 0; i < args.size() + 3; i++)
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl_fused_ops_test.cc]: Nit: we prefer `++i` to `i++`.

File: tensorflow/core/kernels/mkl_fused_ops_test.cc [hunk 5]
```diff
@@ -401,5 +403,262 @@ TEST_F(FusedPadConvOpTest, PaddingConvTestNchw) {
...
+    DataType dtype = DataTypeToEnum<T>::v();
+    const int num_args = static_cast<int>(args.size());
+    Tensor padding(DT_INT32, {4, 2});
+    test::FillValues<int32>(&padding, {0, 0, padding_list[0], padding_list[1],
+                                       padding_list[2], padding_list[3], 0, 0});
+
+    TF_EXPECT_OK(NodeDefBuilder("pad_fused_conv_op", "_MklPadWithFusedConv2D")
+                     .Input(FakeInput(dtype))
+                     .Input(FakeInput(dtype))
+                     .Attr("num_args", num_args)
```
[the remark this answers is not in the document]
Comment (penpornk) [reply tensorflow/core/kernels/mkl_fused_ops_test.cc]: Sorry for being ambiguous. I meant let's keep `Input` and `Attr` in a similar order to when they are defined in `REGISTER_OP`:
https://github.com/tensorflow/tensorflow/blob/34d2f298f78d009b5ad2732e436cdae640509a33/tensorflow/core/ops/mkl_nn_ops.cc#L88-L111

So in this case I was talking about moving `num_args` to between `T` and `strides`.
```c++
    TF_EXPECT_OK(NodeDefBuilder("pad_fused_conv_op", "_MklPadWithFusedConv2D")
                     .Input(FakeInput(dtype))
                     .Input(FakeInput(dtype))
                     .Input(FakeInput(DT_UINT8))
                     .Input(FakeInput(num_args, DT_UINT8))
                     .Input(FakeInput(DT_UINT8))
                     .Attr("T", dtype)
                     .Attr("num_args", num_args)
                     .Attr("strides", {1, stride, stride, 1})
                     .Attr("padding", "VALID")
                     .Attr("fused_ops", fused_ops)
```

**Review by penpornk — CHANGES_REQUESTED** [review]
No worries! Sorry for my delay as well. Thank you so much for the changes. I think we should be good to go after a few very minor changes.

**Comment by guizili0** [comment]
@penpornk thanks for you update, the changes you requested had been addressed, please help to review.

**Review by penpornk — DISMISSED** [review]
Thank you very much again for the PR! :)

**Comment by guizili0** [comment]
@penpornk Can you help to check these failure, it seems did not related to my code, and others PR also had failure in these UT.

**Comment by agramesh1** [comment]
pinging @caisq also

**Comment by caisq** [comment]
@agramesh1 Roger. Starting the testing and merging process.

**Comment by penpornk** [comment]
@guizili0 @agramesh1 Sorry for my delay! Thank you, @caisq !

**Comment by caisq** [comment]
Experiencing some glitches in the sync'ing mechanisms. Stay tuned. cc @yifeif

**Comment by yifeif** [comment]
sync should be resolved, thanks for the patience!

**Comment by yifeif** [comment]
@guizili0 do you mind resolving the merge conflicts before we proceed? Thank you!

**Comment by guizili0** [comment]
@yifeif conflict has been resolved, thanks.

**Comment by penpornk** [comment]
For some reason, I'm getting a message that there is a conflict again. @guizili0, could you please help take a look? Sorry for the inconvenience!

**Comment by guizili0** [comment]
@penpornk I cannot find the conflict, and the merge status is "Merging can be performed automatically with 1 approving review." Can you help to check it?

**Comment by penpornk** [comment]
@guizili0 It's getting merged now. Sorry for the trouble!

**Pull request merged**